### ISSUE TITLE
Just loading

### ISSUE THREAD

**Issue by VladoPortos** [sheyabernstein/MMM-pihole-stats#1]
Hello,
Not sure if this is the correct way to get help, but I'll try :)

I setup the module as specified in wiki

It looks like this in config:
`                {
                        module: 'MMM-pihole-stats',
                        position: 'top_left',
                        config: {
                                apiUrl: 'http://pi.hole/admin/api.php',
                                showSources: true
                                }
                },
`

The /etc/lighttpd/lighttpd.conf looks like this

`# If the URL starts with /admin, it is the Web interface
$HTTP["url"] =~ "^/admin/" {
        # Create a response header for debugging using curl -I
    setenv.add-response-header = (
        "X-Pi-hole" => "The Pi-hole Web interface is working!",
        "X-Frame-Options" => "DENY",
        "Access-Control-Allow-Origin" => "*",
    )
}
`

From the console I can visit the domain http://pi.hole/admin/api.php and get:
`{"domains_being_blocked":"104,575","dns_queries_today":"3,942","ads_blocked_today":"319","ads_percentage_today":"8.1"}`

So that is working... but in the MMM it self its just saying **"Loading..."** 

Any idea what might be wrong ?

**Comment by sheyabernstein** [comment]
What errors come up when debugging in a browser?

**Comment by VladoPortos** [comment]
Hmm I'm not sure how to do it, pretty much started with MMM today :D if you tell me how to check I will 👍

**Comment by VladoPortos** [comment]
I noticed in error log that every time MMM starts it show this:

016-12-26 23:18:16: Failed CORS: http://localhost:8080 vs http://10.0.0.176, http://pi-hole, http://pi.hole, http://localhost

**Comment by sheyabernstein** [comment]
Try restarting the Pi-Hole server by running `sudo service lighttpd restart`

**Comment by VladoPortos** [comment]
Tried that many times, but no luck. Its midnight here so I'm off for some sleep, will continue to look at at it later. My suspicion is something with that  "Failed CORS" maybe it won't allow the script to pull the data...  not sure.

**Comment by sheyabernstein** [comment]
Try enabling `mod_setenv`
http://www.gubatron.com/blog/2011/11/29/lighttpd-allow-access-control-allow-origin-headers-on-the-server-status-page/

**Comment by VladoPortos** [comment]
Hello, 
Already did that ( I was googling that error a lot :D but there is not much info about it ) 

My mods are as follow:

```
server.modules = (
        "mod_access",
        "mod_accesslog",
        "mod_auth",
        "mod_expire",
        "mod_compress",
        "mod_redirect",
        "mod_setenv",
        "mod_rewrite",
        "mod_alias"
)

```

**Comment by sheyabernstein** [comment]
Is the mirror running on the same pi as Pi-Hole? Try localhost in the url

**Comment by VladoPortos** [comment]
Yes it is running from the same Pi

Just tried:
apiUrl: 'http://127.0.0.1/admin/api.php',
and
apiUrl: 'apiUrl: 'http://127.0.0.1/admin/api.php',',
and
apiUrl: 'http://pi.hole/admin/api.php',

But no luck still stuck on loading. The funny thing is when I visit each of the above links via links2 from console they shows the correct info.

**Comment by D3rpi** [comment]
@VladoPortos Did you ever  get it to work? I am having the same issue here.

**Comment by xadacka** [comment]
Followed this and forgot all about it, but also interested in a fix. Developer console is telling me there's a syntax error but I can't figure out what it is, errors are:

```
VM235:1 Uncaught SyntaxError: Unexpected token F in JSON at position 0
    at JSON.parse (<anonymous>)
    at XMLHttpRequest.statsSourcesRequest.onreadystatechange (MMM-pihole-stats.js:144)

Uncaught SyntaxError: Unexpected token F in JSON at position 0
    at JSON.parse (<anonymous>)
    at XMLHttpRequest.statsSummaryRequest.onreadystatechange (MMM-pihole-stats.js:120)
```

Line 144 is: 
`self.processSources(JSON.parse(this.response));`

Line 120 is: 
`self.processSummary(JSON.parse(this.response));`

**Comment by VladoPortos** [comment]
@D3rpi Sadly didn't and haven't had time to investigate more yet, so skipped it for now.

**Comment by mckennajones** [comment]
@sheyabernstein any update on this. Having the same issue. I would love to use this module!

**Comment by mckennajones** [comment]
The specific error I'm getting is `XMLHttpRequest cannot load http://localhost/admin/api.php?summary. The 'Access-Control-Allow_Origin' header contains multiple values '*, http://localhost:8080' but only one is allowed.`

Oddly enough, if I access the api page in a browser, the Access-Control-Allow_Origin only contains *.

**Comment by mckennajones** [comment]
The error that I'm getting is that Access-Control-Allow-Origin has multiple values, * and localhost:8080.

In the readme it says to set Access-Control-Allow-Origin to * in the lighttpd config. I did this, and if I access the page in a browser it returns the correct header, however, with the XMLHttpRequest call in the JS something weird is happening...

**Comment by sheyabernstein** [comment]
@VladoPortos @D3rpi @mckennajones Just updated and tested the code with a fresh install of magic mirror and pi-hole. it should all work without having to change lighttpd configs if the mirror is being viewed on the same machine as the pi-hole is on. Change the `apiURL` in the configs for MMM-pihole-stats to `http://<pi's ip address>/admin/api.php` if on a different machine.

My current `~/MagicMirror/config/config.js`:

```
'''
 {
     module: 'MMM-pihole-stats',
     position: 'top_left',
     config: {
         'http://192.168.1.38/admin/api.php'
     }
 }
...
```

I hope it works for you 😄

**Comment by mckennajones** [comment]
@sheyabernstein Cheers! Thanks. Will try it later tonight.

**Comment by VladoPortos** [comment]
@sheyabernstein thanks, tested it just now, but still stuck on "Loading..." can you post your lighttpd.conf maybe I have something wrong there...

**Comment by sheyabernstein** [comment]
@VladoPortos this is my current lighttpd.conf in `/etc/lighttpd/lighttpd.conf`. It's the default pi-hole conf.
```
# Pi-hole: A black hole for Internet advertisements
# (c) 2015, 2016 by Jacob Salmela
# Network-wide ad blocking via your Raspberry Pi
# http://pi-hole.net
# lighttpd config for Pi-hole
#
# Pi-hole is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 2 of the License, or
# (at your option) any later version.

server.modules = (
	"mod_access",
	"mod_accesslog",
	"mod_auth",
	"mod_expire",
	"mod_compress",
	"mod_redirect",
	"mod_setenv",
	"mod_rewrite"
)

server.document-root        = "/var/www/html"
server.error-handler-404	= "pihole/index.php"
server.upload-dirs          = ( "/var/cache/lighttpd/uploads" )
server.errorlog             = "/var/log/lighttpd/error.log"
server.pid-file             = "/var/run/lighttpd.pid"
server.username             = "www-data"
server.groupname            = "www-data"
server.port                 = 80
accesslog.filename			= "/var/log/lighttpd/access.log"
accesslog.format			= "%{%s}t|%V|%r|%s|%b"


index-file.names            = ( "index.php", "index.html", "index.lighttpd.html" )
url.access-deny             = ( "~", ".inc" )
static-file.exclude-extensions = ( ".php", ".pl", ".fcgi" )

compress.cache-dir          = "/var/cache/lighttpd/compress/"
compress.filetype           = ( "application/javascript", "text/css", "text/html", "text/plain" )

# default listening port for IPv6 falls back to the IPv4 port
include_shell "/usr/share/lighttpd/use-ipv6.pl " + server.port
include_shell "/usr/share/lighttpd/create-mime.assign.pl"
include_shell "/usr/share/lighttpd/include-conf-enabled.pl"

# If the URL starts with /admin, it is the Web interface
$HTTP["url"] =~ "^/admin/" {
	# Create a response header for debugging using curl -I
    setenv.add-response-header = (
        "X-Pi-hole" => "The Pi-hole Web interface is working!",
        "X-Frame-Options" => "DENY"
    )
}

# Rewite js requests, must be out of $HTTP block due to bug #2526
url.rewrite = ( "^(?!/admin/).*\.js$"  => "pihole/index.js"   )

# If the URL does not start with /admin, then it is a query for an ad domain
$HTTP["url"] =~ "^(?!/admin)/.*" {
    # Create a response header for debugging using curl -I
    setenv.add-response-header = ( "X-Pi-hole" => "A black hole for Internet advertisements." )
}

# Add user chosen options held in external file
include_shell "cat external.conf 2>/dev/null"
```

**Comment by VladoPortos** [comment]
@sheyabernstein hmm have the same, no idea why its not working :(

**Comment by sheyabernstein** [comment]
@VladoPortos is it still the no access control allow origin error?

**Comment by D3rpi** [comment]
@sheyabernstein It's working now!

**Comment by endofzero** [comment]
I'm seeing the same behavior. Electron.js will pretty much cap all threads when this is enabled.

Pi-hole is on a different IP than the MM, and I can get the json string without issue using a normal browser.

With the module enabled, my 5m load average shot up to 4.16.

**Comment by sheyabernstein** [comment]
@endofzero the load average shooting up is because of the multiple PHP threads being created - pih-hole doesn't limit them properly. What error are you getting in the MM console?

**Comment by endofzero** [comment]
I'll provide the details when I'm able to test.

However, I don't understand your statement about pi-hole not limiting them correctly. As I had stated, pi-hole is not on the same machine as MM, how is the query of the pi-hole json affecting php threads in MM to cause an increase in cpu? Seems that would still be an issue with the plugin, not pi-hole, with them being on different PIs

**Comment by endofzero** [comment]
It looks like my settings are being ignored and the defaults are getting used instead. So it's the same as what was reported previously.

Here is the configuration setup

{
     module: 'MMM-pihole-stats',
     position: 'top_left', // Or any valid MagicMirror position.
     config: {
        apiUrl: 'http://192.168.1.119:81/admin/api.php',
     }
 },


Error:
XMLHttpRequest cannot load http://pi.hole/admin/api.php?getQuerySources. No 'Access-Control-Allow-Origin' header is present on the requested resource. Origin 'http://192.168.1.130:8080' is therefore not allowed access. The response had HTTP status code 404.

I was able to get access ONLY by hardcoding my value in the actual file. Seems the defaults are overwritting the settings in the config file.

**Comment by pic0frame** [comment]
I had the same issue. The problem is that 'apiURL' variable. So to fix change 'apiUrl' to 'apiURL' in config.

**Comment by sheyabernstein** [comment]
@pic0frame changed all to apiURL

**Comment by flo80** [comment]
I had the same no 'Access-Control-Allow-Origin' issue, the reason is within pi-hole. The admin scripts check for cross-site requests, I work around it by allowing my second host as well in the check.

In file _/var/www/html/admin/scripts/pi-hole/php/auth.php_ add your host in the array of allow hosts:

`
    $AUTHORIZED_HOSTNAMES = array(
        $ipv4,
        $ipv6,
        str_replace(array("[","]"), array("",""), $_SERVER["SERVER_NAME"]),
        "pi.hole",
        "localhost",
        "your.hostname.com"
    );`

This will most likely break with the next pi-hole update but is pretty easy fix again.

**Comment by E3V3A** [comment]
Perhaps contact the pi-hole maintainer, to permanently resolve this?

**Comment by marco-hofmann** [comment]
DId you ever get it to work? I've got the same problem

**Comment by TATUMTOT** [comment]
Has this Died?

**Comment by nickosouras** [comment]
I think so sadly

**Comment by maximilianfriedrich** [comment]
I had this problem always when I had something different than `var config = {
	address: "localhost", ...` in `config/config.js`.
This is why I can't use MMM-Remote-Control and the MMM-pihole-stats at the same time.

Hope this helps.

**Comment by nickosouras** [comment]
what was the address normally? Also does the pihole have to be hosted on the same device as the magicmirror?

**Comment by maximilianfriedrich** [comment]
`address: "localhost",` is default, I'm still trying to get this running with `address: "0.0.0.0",` or something like this, to enable access from other local devices as my laptop.

Pi-Hole itself is running on a different Raspberry within my LAN, so I access it with `apiURL: "http://192.168.XXX.XXX/admin/api.php"` from my MMM.

Like I said, playing around with the address solved it partially, but I'm not yet happy with this solution. What worries me a little bit more is the MMM-pihole-stats module seams to cause to freezes/crashs of the MMM.

Still not happy with this module but I will keep trying to get it running smoothly.

**Comment by nickosouras** [comment]
Ill have to play around with the configuration and see if I can get it working. Hopefully someone tries to rewrite it so it isn't busted anymore but heres to hoping. thanks for the replies

**Comment by sheyabernstein** [comment]
I haven't used this module in a while, but I remember there were sometimes issus with `showSources` . Setting it to `false` in the config worked for me.

**Comment by Barokai** [comment]
i encountered the same problems today (installed the module the first time), i had 2 things to change to get rid of the "LOADING..."

First:
editing (maybe empty)  /etc/lighttpd/external.conf (edit this instead of /etc/lighttpd/lighttpd.conf to avoid your changes being overridden on pihole updates)
```
# If the URL starts with /admin, it is the Web interface
$HTTP["url"] =~ "^/admin/" {
    $HTTP["url"] =~ "api.php" {
        setenv.add-response-header = ( "Access-Control-Allow-Origin" => "*" )
    }
} 
```
and restarting lighthttpd aftewards with `sudo service lighttpd restart`

Second:
disabline websecurity for electron as found here: 
https://forum.magicmirror.builders/post/48208
```
electronOptions: {
    webPreferences: {
        webSecurity: false
    }
}
```
in short: placing this before or after the modules section in ~/MagicMirror/config/config.js

**Comment by jhetland** [comment]
For me it was enough just to add this line in the MagicMirror config:
` electronOptions: {
    webPreferences: {
        webSecurity: false
    }`

My guess is that this is because pi.hole is only available over http locally OOTB, and there is an expectation for https by MagicMirror. So by allowing MagicMirror to talk insecurly, it works. 

Jo

**Comment by halolly** [comment]
thx jhetland, that was the soultion for me
electronOptions: { webPreferences: { webSecurity: false }},

although... that breaks my netamo module which doesn´t show up anymore.... argh

**Comment by Plainbroke** [comment]
Out of 26 modules running on my MagicMirror,  This is the only one I can not get to function.  It either says "loading" or freezes my Pi4.    HELP HELP HELP.

**Comment by sheyabernstein** [comment]
This is been fixed in the latest commit. Please check out item 3 in the readme

**Comment by Plainbroke** [comment]
Hey Sheya,
I am still freezing up.
Here is what I put in the external.conf file.  My Pihole did not have the external.conf  file so I made the external.conf file then 
Added this in it.

HTTP["192.168.100.9"] =~ "^/admin/api.php" {         ///This is the Pihole IP
  setenv.add-environment = (
                "SERVER_NAME" => "{192.168.100.222}"   /// This is the MagicMirrors IP
        )
}

Both work fine by them selves. But when I try your module it freezes MagicMirror up 3-5 minutes of running

Jeff Burdick
